packages/arb-token-bridge-ui/src/hooks/useCompleteMultiChainTransactions.ts

```diff
@@ -0,0 +1,339 @@
+        })
+      )
+    }
+  )
+
+  useEffect(() => {
+    if (data) {
+      // if there is not a single full page in any of the data fetched, then there is no more to fetch
+      const shouldFetchNextPage = data.some((_, index) =>
+        shouldFetchPage(index)
+      )
+
+      // include the new data with the previously fetched data
+      // the data is grouped by chain pairs
+      setTransactions(prevTransactions => {
+        return data.map((transactionsForChainPair, chainPairIndex) => {
+          const prevTransactionsForChainPair = prevTransactions[chainPairIndex]
+          return [
+            ...(prevTransactionsForChainPair
+              ? prevTransactionsForChainPair
+              : []),
+            ...(transactionsForChainPair ? transactionsForChainPair : [])
+          ]
+        })
+      })
+
+      if (!shouldFetchNextPage) {
+        setLoading(false)
+        return
+      }
+
+      setPage(prevPage => prevPage + 1)
+    }
+  }, [data, shouldFetchPage])
+
+  return {
+    data: transactions.flat(),
+    loading,
+    error
+  }
+}
+
+/**
+ * Fetches transaction history only, without mapping additional info that would require a lot of RPC calls.
+ * The data is collected for multiple chain pairs, and sorted by date.
+ */
+const useMultiChainTransactionList = () => {
+  const deposits = useTransactionListByDirection('deposits')
+  const withdrawals = useTransactionListByDirection('withdrawals')
+
+  const groupedTransactions = [deposits, withdrawals]
+
+  // merge grouped transactions and sort them by date
+  const data = groupedTransactions
+    .flatMap(grp => grp.data)
+    .sort(sortByTimestampDescending)
+  // checks if any source is loading
+  const loading = groupedTransactions.map(grp => grp.loading).some(Boolean)
+  // get first error
+  const error =
+    groupedTransactions.map(grp => grp.error).filter(Boolean)[0] ?? null
+
+  return { data, loading, error }
+}
+
+/**
+ * Maps additional info to previously fetches transaction history, starting with the earliest data.
+ * This is done in small batches to safely meet RPC limits.
+ */
+export const useCompleteMultiChainTransactions = () => {
+  // max number of transactions mapped in parallel, for the same chain pair
+  // we can batch more than MAX_BATCH_SIZE at a time if they use a different RPC
+  // MAX_BATCH_SIZE means max number of transactions in a batch for a chain pair
+  const MAX_BATCH_SIZE = 5
+
+  const [transactions, setTransactions] = useState<MergedTransaction[]>([])
+  const [page, setPage] = useState(0)
+  const [fetching, setFetching] = useState(true)
+  const [paused, setPaused] = useState(false)
+
+  const { data, loading, error } = useMultiChainTransactionList()
+  const { address } = useAccount()
+
+  const { data: mapData, error: mapError } = useSWRImmutable(
+    address && !loading ? [address, page] : null,
+    ([, _page]) => {
+      // TODO: Need to allow more than MAX_BATCH_SIZE if they are for diff chain pairs
```

Comment:
i think we should always pair a TODO with a github issue ticket

so it depends on whether this TODO will be done before launching the new tx history

if not, we should create a github ticket and add it to this comment

Reply:
it will be done before, it's just an optimization note to myself